src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
+        private readonly Logger logger;
+        private TStage highStage;
+
+        public LifecycleObservable(Logger logger)
+        {
+            this.logger = logger?.GetLogger(GetType().Name);
+            subscribers = new ConcurrentDictionary<object, OrderedObserver>();
+        }
+
+        public async Task OnStart()
+        {
+            try
+            {
+                foreach (IGrouping<TStage, OrderedObserver> observerGroup in subscribers.Values
```

Comment:
I agreed that it could be optimized, and that in extreme cases it could be costly.  I also agree that it may be valuable to revisit this in the future and introduce a grain specific lifecycle (it is an abstraction after all) which is more tuned to the performance needs of grain.  I do not agree that grain activation is part of the hot path.  Given all that grain activation does, the number of observers would need to be absurdly high before this ever even showed up as a blip in profiling.  Spending time optimizing this now, before we've any reason to believe it will effect performance at all, is, imo, premature optimization.